kai_mcp_solution_server/src/kai_mcp_solution_server/server.py

```
@@ -36,6 +39,54 @@
 
 
+def with_db_recovery(func):
+    """Decorator to execute database operations with automatic recovery on connection errors.
+
+    Uses a semaphore to limit concurrent DB operations and prevent pool exhaustion.
+    Implements exponential backoff with retry on connection errors.
+    """
+
+    async def wrapper(kai_ctx: KaiSolutionServerContext, *args, **kwargs):
+        if _SharedResources.db_semaphore is None:
+            raise RuntimeError("Database semaphore not initialized")
```

Ruff (0.13.1)

Avoid specifying long messages outside the exception class

(TRY003)